### DISCUSSION TITLE
Event listener is sometimes not called after the event is fired

### DISCUSSION THREAD

**Discussion by allthingsliquid** [socketio/socket.io#3996] [Q&A] [answered]
Hi, 
I recently upgraded socket.io from 2.3.0 to 4.1.2 and encountered some issues with listeners.

This is a sample of the code I used: 

```typescript
function(socket: SocketIO.Socket) {
  ConnectionCounter.increment()
  logConnection(socket)

  socket.use(SocketMiddlewares.logEvent(socket))
  socket.on('disconnect', SocketEventHandlers.onDisconnect)

  // The pipe function allows us to chain multiple function calls
  socket.on('stopMachine', pipe(fixEnvForMsg, SocketEventHandlers.onStopMachine))
  // A bunch more events like this below
  // ...
}
```

I have a confirmation (through log lines) that the event is fired and that the machine receives it. But it seems that the listener isn't always called after that.

It only happens sometimes though, and I can't find an explanation to it (about 10 times out of 200)
 
A few more things to note:
- Nothing has changed in the `pipe` function nor in the chained functions inside before and after the upgrade 
- It does not depend on the event nor the sent data, and usually resending the same event with the same payload will end up calling the listener 
- When it happens, the listener doesn't even seem to be called at all (log lines that should've appeared if called are not visible)

Was there any change on the `socket.on` function? 
Thank you in advance for your help!

**Comment by darrachequesne** [comment, marked as answer]
Hi! That seems weird indeed. The only change that might be related to this behavior is https://github.com/socketio/socket.io/commit/494c64e44f645cbd24c645f1186d203789e84af0, but I'm not sure that's your case.

Are you able to reliably reproduce? A sample project reproducing the issue would help a lot.

**Reply by allthingsliquid** [reply]
Hi!

I think it could totally be related to this. The client which fires the event closes immediately after. I guess it is possible that sometimes the server receives the event after the client connection is closed, thus not reacting on the emitted event.

I can't provide a sample project, but I'll test if keeping the client connection open fixes the problem and will keep you updated.

Thank you for your help!
